✨ CAPD and E2E framework tests failure domains (#6178)
kubernetes-sigs/cluster-api: docs/book/src/developer/providers/v1.1-to-v1.2.md

### Other

> - Following E2E framework functions are now checking for machine being created in the expected failure domain (if defined);

Reviewer: ```suggestion
- Following E2E framework functions are now checking that machines are created in the expected failure domain (if defined);
```

>   as a consequence, all the E2E test can now exercise failure domain too.

Reviewer: ```suggestion
  as a consequence, all E2E tests can now verify failure domains too.
```
nit. Not sure if that verb is better, exercise just sounded strange to me.

Feel free to ignore

  - `ApplyClusterTemplateAndWait`
  - `WaitForControlPlaneAndMachinesReady`
  - `DiscoveryAndWaitForMachineDeployments`

> - The `AssertControlPlaneFailureDomains` in the E2E has been modified to allow proper failure domain testing

Reviewer: ```suggestion
- The `AssertControlPlaneFailureDomains` function in the E2E test framework has been modified to allow proper failure domain testing.
```
